build_tools/pexpect_helper.py

```diff
@@ -0,0 +1,261 @@
+
+    def formatted(self):
+        """ Return a human-readable string representing this message. """
+        etext = escape(self.text)
+        timestamp = self.when * 1000.0
+        return "{dir} {timestamp:.2f} ({filename}:{lineno}): {etext}".format(
+            timestamp=timestamp, etext=etext, **vars(self)
+        )
+
+
+class SpawnedProc(object):
+    """ A process, talking to our ptty. This wraps pexpect.spawn.
+
+    Attributes:
+        colorize: whether error messages should have ANSI color escapes
+        messages: list of Message sent and received, in-order
+        start_time: the timestamp of the first message, or None if none yet
+        spawn: the pexpect.spawn value
+        prompt_counter: the index of the prompt. This cooperates with the fish_prompt
+            function to ensure that each printed prompt is distinct.
+    """
+
+    def __init__(self, name="fish", timeout=TIMEOUT_SECS, env=os.environ.copy()):
+        """ Construct from a name, timeout, and environment.
+
+            Args:
+                name: the name of the executable to launch, as a key into the
+                      environment dictionary. By default this is 'fish' but may be
+                      other executables.
+                timeout: A timeout to pass to pexpect. This indicates how long to wait
+                         before giving up on some expected output.
+                env: a string->string dictionary, describing the environment variables.
+        """
+        if name not in env:
+            raise ValueError("'name' variable not found in environment" % name)
+        exe_path = env.get(name)
+        self.colorize = sys.stdout.isatty()
+        self.messages = []
+        self.start_time = None
+        self.spawn = pexpect.spawn(exe_path, env=env, encoding="utf-8", timeout=timeout)
+        self.spawn.delaybeforesend = None
+        self.prompt_counter = 1
+
+    def time_since_first_message(self):
+        """ Return a delta in seconds since the first message, or 0 if this is the first. """
+        now = time.monotonic()
+        if not self.start_time:
+            self.start_time = now
+        return now - self.start_time
+
+    def send(self, s):
+        """ Cover over pexpect.spawn.send().
+            Send the given string to the tty, returning the number of bytes written.
+        """
+        res = self.spawn.send(s)
+        when = self.time_since_first_message()
+        self.messages.append(Message.sent(s, when))
+        return res
+
+    def sendline(self, s):
+        """ Cover over pexpect.spawn.sendline().
+            Send the given string + linesep to the tty, returning the number of bytes written.
+        """
+        return self.send(s + os.linesep)
+
+    def expect_re(self, pat, pat_desc=None, unmatched=None, **kwargs):
+        """ Cover over pexpect.spawn.expect().
+            Look through the "new" output of self.spawn until the given pattern is matched.
```

Comment:
Can we add to the docstring that this consumes the output until after the match? Kind of obvious, but useful to someone new to expect.

Reply:
Sure, good suggestion!